q2_feature_table/_filter.py

```diff
@@ -6,16 +6,116 @@
+    # 2) The substitution syntax recommended in the docs doesn't allow
+    #    us to specify complex where statements, which is what we need to do
+    #    here. For example, we need to specify things like:
+    #        WHERE Subject='subject-1' AND SampleType='gut'
+    #    but their qmark/named-style syntaxes only supports substition of
+    #    variables, such as:
+    #        WHERE Subject=?
+    # 3) sqlite3.Cursor.execute will only execute a single statement so
+    #    inserting multiple statements (e.g., "Subject='subject-1'; DROP...")
+    #    will result in an OperationalError being raised.
+    query = ('SELECT "{0}" FROM metadata WHERE {1} GROUP BY "{0}" '
+             'ORDER BY "{0}";'.format(id_column, where))
+
+    try:
+        c.execute(query)
+    except sqlite3.OperationalError:
+        conn.close()
+        raise ValueError("Selection of ids failed with query:\n %s"
+                         % query)
+
+    ids = c.fetchall()
+    conn.close()
+    return ids
 
 
-def filter(table: biom.Table, tree: skbio.TreeNode) -> biom.Table:
-    """ Filter feature table in-place against a phylogenic tree
-    """
-    tip_ids = set([t.name for t in tree.tips()])
-    feature_ids = set(table.ids(axis='observation'))
-    # ids_to_keep can only include ids that are in table
-    ids_to_keep = tip_ids & feature_ids
-    table.filter(ids_to_keep, axis='observation', inplace=True)
+def _get_biom_filter_function(ids_to_keep, min_count, max_count,
+                              min_nonzero, max_nonzero):
+    ids_to_keep = set(ids_to_keep)
+    if max_count is None:
+        max_count = np.inf
+    if max_nonzero is None:
+        max_nonzero = np.inf
+
+    def f(data_vector, id_, metadata):
+        return (id_ in ids_to_keep) and \
+               (min_count <= data_vector.sum() <= max_count) and \
+               (min_nonzero <= (data_vector > 0).sum() <= max_nonzero)
+    return f
+
+
+def _filter(table, min_count, max_count, min_nonzero, max_nonzero, metadata,
+            where, axis):
+    if min_count == 0 and max_count is None and min_nonzero == 0 and\
+       max_nonzero is None and metadata is None and where is None:
+        raise ValueError("No filtering was requested.")
+    if metadata is None and where is not None:
+        raise ValueError("Metadata must be provided if where is "
+                         "specified.")
+
+    if where is not None:
+        ids_to_keep = _ids_where(metadata, where)
+    elif metadata is not None:
+        ids_to_keep = metadata.to_dataframe().index
+    else:
+        ids_to_keep = table.ids(axis=axis)
+
+    filter_fn = _get_biom_filter_function(ids_to_keep, min_count, max_count,
+                                          min_nonzero, max_nonzero)
+    table.filter(filter_fn, axis=axis, inplace=True)
     return table
+
+
+def filter_samples(table: biom.Table, min_count: int=0, max_count: int=None,
+                   min_features: int=0, max_features: int=None,
```

Comment:
`min_features` could also be called `min_nonzero`, since it's the number of nonzero values that must be in a sample vector. I think `min_features` is more clear, but I wanted to throw out the alternative of calling this `min_nonzero` (in which case `max_features` would become `max_nonzero`).